libs/utils/src/properties_encoding.c

```diff
@@ -0,0 +1,580 @@
+    } else if (json_is_object(jsonValue)) {
+        const char* fieldName;
+        json_t* fieldValue;
+        json_object_foreach(jsonValue, fieldName, fieldValue) {
+            char buf[64];
+            char* combinedKey = celix_utils_writeOrCreateString(buf, sizeof(buf), "%s/%s", key, fieldName);
+            celix_auto(celix_utils_string_guard_t) strGuard = celix_utils_stringGuard_init(buf, combinedKey);
+            if (!combinedKey) {
+                celix_err_push("Failed to create sub key.");
+                return ENOMEM;
+            }
+            status = celix_properties_decodeValue(props, combinedKey, fieldValue, flags);
+            if (status != CELIX_SUCCESS) {
+                return status;
+            }
+        }
+        return CELIX_SUCCESS;
+    } else if (json_is_array(jsonValue) && json_array_size(jsonValue) == 0) {
+        if (flags & CELIX_PROPERTIES_DECODE_ERROR_ON_EMPTY_ARRAYS) {
+            celix_err_pushf("Invalid empty array for key '%s'.", key);
+            return CELIX_ILLEGAL_ARGUMENT;
+        }
+        // ignore empty arrays
+        return CELIX_SUCCESS;
+    } else if (json_is_array(jsonValue)) {
+        status = celix_properties_decodeArray(props, key, jsonValue, flags);
+    } else if (json_is_null(jsonValue)) {
+        if (flags & CELIX_PROPERTIES_DECODE_ERROR_ON_NULL_VALUES) {
+            celix_err_pushf("Invalid null value for key '%s'.", key);
+            return CELIX_ILLEGAL_ARGUMENT;
+        }
+        // ignore null values
+        return CELIX_SUCCESS;
+    } else {
+        // LCOV_EXCL_START
+        celix_err_pushf("Unexpected json value type for key '%s'.", key);
+        return CELIX_ILLEGAL_ARGUMENT;
+        // LCOV_EXCL_STOP
+    }
+    return status;
+}
+
+static celix_status_t celix_properties_decodeFromJson(json_t* obj, int flags, celix_properties_t** out) {
+    *out = NULL;
+    if (!json_is_object(obj)) {
+        celix_err_push("Expected json object.");
+        return CELIX_ILLEGAL_ARGUMENT;
+    }
+
+    celix_autoptr(celix_properties_t) props = celix_properties_create();
+    if (!props) {
+        return ENOMEM;
+    }
+
+    const char* key;
+    json_t* value;
+    json_object_foreach(obj, key, value) {
+        celix_status_t status = celix_properties_decodeValue(props, key, value, flags);
+        if (status != CELIX_SUCCESS) {
+            return status;
+        }
+    }
+
+    *out = celix_steal_ptr(props);
+    return CELIX_SUCCESS;
+}
+
+celix_status_t celix_properties_loadFromStream(FILE* stream, int decodeFlags, celix_properties_t** out) {
+    json_error_t jsonError;
+    size_t jsonFlags = 0;
+    if (decodeFlags & CELIX_PROPERTIES_DECODE_ERROR_ON_DUPLICATES) {
+        jsonFlags = JSON_REJECT_DUPLICATES;
+    }
+    json_auto_t* root = json_loadf(stream, jsonFlags, &jsonError);
+    if (!root) {
+        celix_err_pushf("Failed to parse json: %s.", jsonError.text);
+        return CELIX_ILLEGAL_ARGUMENT;
```

Comment:
`CELIX_ILLEGAL_ARGUMENT` may be completely misleading, because `json_error_code` might return the following errors:

```C
enum json_error_code {
    json_error_unknown,
    json_error_out_of_memory,
    json_error_stack_overflow,
    json_error_cannot_open_file,
    json_error_invalid_argument,
    json_error_invalid_utf8,
    json_error_premature_end_of_input,
    json_error_end_of_input_expected,
    json_error_invalid_syntax,
    json_error_invalid_format,
    json_error_wrong_type,
    json_error_null_character,
    json_error_null_value,
    json_error_null_byte_in_key,
    json_error_duplicate_key,
    json_error_numeric_overflow,
    json_error_item_not_found,
    json_error_index_out_of_range
};
```

I suggest some error code remapping here.